File: BSTIterator.cpp

#include<deque>
#include<vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}
};

class BSTIterator {
    
public:

    deque<TreeNode*> nextNodes;
    vector<TreeNode*> previosNodes;
    int indexPrevious = -1;

    BSTIterator(TreeNode* root) {
        iterateToNextLeftmostNodeFromCurrentRoot(root);
    }

    bool hasNext() {
        return !nextNodes.empty() || indexPrevious + 1 < previosNodes.size();
    }

    int next() {
        if (!hasNext()) {
            throw out_of_range("No Such Element Exception.");
        }

        if (indexPrevious + 1 < previosNodes.size()) {
            return previosNodes[++indexPrevious]->val;
        }

        TreeNode* nextNode = nextNodes.front();
        nextNodes.pop_front();
        iterateToNextLeftmostNodeFromCurrentRoot(nextNode->right);

        previosNodes.push_back(nextNode);
        indexPrevious++;

        return nextNode->val;
    }

    bool hasPrev() {
        return indexPrevious - 1 >= 0;
    }

    int prev() {
        if (!hasPrev()) {
            throw out_of_range("No Such Element Exception.");
        }
        return previosNodes[--indexPrevious]->val;
    }

    void iterateToNextLeftmostNodeFromCurrentRoot(TreeNode* node) {
        while (node != nullptr) {
            nextNodes.push_front(node);
            node = node->left;
        }
    }
};

File: BSTIterator.js

function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
}

/**
 * @param {TreeNode} root
 */
var BSTIterator = function (root) {
    this.indexPrevious = -1;
    this.nextNodes = [];
    this.previosNodes = [];
    this.iterateToNextLeftmostNodeFromCurrentRoot(root);
};

/**
 * @return {boolean}
 */
BSTIterator.prototype.hasNext = function () {
    return this.nextNodes.length > 0 || this.indexPrevious + 1 < this.previosNodes.length;
};

/**
 * @return {number}
 */
BSTIterator.prototype.next = function () {
    if (!this.hasNext()) {
        throw "No Such Element Exception.";
    }

    if (this.indexPrevious + 1 < this.previosNodes.length) {
        return this.previosNodes[++this.indexPrevious].val;
    }

    let nextNode = this.nextNodes.pop();
    this.iterateToNextLeftmostNodeFromCurrentRoot(nextNode.right);

    this.previosNodes.push(nextNode);
    this.indexPrevious++;

    return nextNode.val;
};

/**
 * @return {boolean}
 */
BSTIterator.prototype.hasPrev = function () {
    return this.indexPrevious - 1 >= 0;
};

/**
 * @return {number}
 */
BSTIterator.prototype.prev = function () {
    if (!this.hasPrev()) {
        throw "No Such Element Exception.";
    }
    return this.previosNodes[--this.indexPrevious].val;
};

/**
 * @param {TreeNode} node
 */
BSTIterator.prototype.iterateToNextLeftmostNodeFromCurrentRoot = function (node) {
    while (node !== null) {
        this.nextNodes.push(node);
        node = node.left;
    }
};
